src/v/cluster/topic_table.cc

```diff
@@ -791,6 +793,456 @@ topic_table::apply(create_non_replicable_topic_cmd cmd, model::offset o) {
     co_return make_error_code(errc::success);
 }
 
+ss::future<>
+topic_table::fill_snapshot(controller_snapshot& controller_snap) const {
+    auto& snap = controller_snap.topics;
+    for (const auto& [ns_tp, md_item] : _topics) {
+        absl::node_hash_map<
+          model::partition_id,
+          controller_snapshot_parts::topics_t::partition_t>
+          partitions;
+        absl::node_hash_map<
+          model::partition_id,
+          controller_snapshot_parts::topics_t::update_t>
+          updates;
+
+        for (const auto& p_as : md_item.get_assignments()) {
+            std::vector<model::broker_shard> replicas;
+            model::ntp ntp(ns_tp.ns, ns_tp.tp, p_as.id);
+            if (auto upd_it = _updates_in_progress.find(ntp);
+                upd_it != _updates_in_progress.end()) {
+                const auto& upd = upd_it->second;
+                updates.emplace(
+                  p_as.id,
+                  controller_snapshot_parts::topics_t::update_t{
+                    .target_assignment = upd.get_target_replicas(),
+                    .state = upd.get_state(),
+                    .revision = upd.get_update_revision(),
+                    .last_cmd_revision = upd.get_last_cmd_revision()});
+
+                replicas = upd.get_previous_replicas();
+            } else {
+                replicas = p_as.replicas;
+            }
+
+            auto p_it = md_item.partitions.find(p_as.id);
+            vassert(
+              p_it != md_item.partitions.end(),
+              "ntp {} must be present in the partition map",
+              ntp);
+
+            partitions.emplace(
+              p_as.id,
+              controller_snapshot_parts::topics_t::partition_t{
+                .group = p_as.group,
+                .replicas = std::move(replicas),
+                .replicas_revisions = p_it->second.replicas_revisions,
+                .last_update_finished_revision
+                = p_it->second.last_update_finished_revision});
+
+            co_await ss::coroutine::maybe_yield();
```

Comment:
are we sure we can yield here ? Is it possible that topic/partition map will change during that operation ?

Reply:
should be okay, we are doing it with `mux_state_machine::_apply_mtx` locked